applications/GeoMechanicsApplication/custom_workflows/adaptive_time_incrementor.cpp

```diff
@@ -73,21 +75,31 @@ double AdaptiveTimeIncrementor::GetIncrement() const { return mDeltaTime; }
 
 void AdaptiveTimeIncrementor::PostTimeStepExecution(const TimeStepEndState& rResultantState)
 {
-    if (rResultantState.NonConverged() ||
-        (rResultantState.Converged() && (rResultantState.num_of_iterations == mMaxNumOfIterations))) {
-        mDeltaTime *= mReductionFactor;
-    } else if (rResultantState.Converged() && (rResultantState.num_of_iterations < mMinNumOfIterations)) {
-        mDeltaTime = std::min(mDeltaTime * mIncreaseFactor, mMaxDeltaTime);
+    if (rResultantState.Converged()) // it is converged also at the beginning of the cycles
+    {
+        if (rResultantState.num_of_iterations < mMinNumOfIterations) {
+            mDeltaTime = std::min(mDeltaTime * mIncreaseFactor, mMaxDeltaTime);
+        } else if (rResultantState.num_of_iterations == mMaxNumOfIterations) {
+            mDeltaTime *= mReductionFactor;
+        }
+        if (rResultantState.time < mEndTime) {
+            // Up-scaling to reach end_time without small increments
+            constexpr auto small_increment_factor = 1.0e-3;
+            if (auto small_time_increment = small_increment_factor * mDeltaTime;
+                rResultantState.time + mDeltaTime > mEndTime - small_time_increment) {
+                mDeltaTime = mEndTime - rResultantState.time;
+            }
```

Comment:
Let's avoid introducing a second minimum time increment. Let's use the minimum time increment that you'll introduce with this PR. For instance:
```suggestion
        if (rResultantState.time < mEndTime && mEndTime - rResultantState.time - mDeltaTime <
                                                   mMaybeMinDeltaTime.value_or(default_min_delta_time)) {
            // Up-scaling to reach end_time without small increment
            mDeltaTime = mEndTime - rResultantState.time;
        }
```

Note that similar changes are required in the Python code.